### PR TITLE
Mandate role claims in gateway-controller when IDP auth is enabled

### PR DESCRIPTION
**Pull request by thivindu**
### Purpose

Improve IDP/JWT auth configuration in the gateway-controller: `roles_claim` had no default, and token audience (aud) wasn't configurable at the controller level.

### Goals

- Default `roles_claim` to "roles" and validate it's non-empty when IDP is enabled.
- Add configurable audience (aud) validation to the controller.

### Approach

- pkg/config: `roles_claim` defaults to "roles"; validateAuthConfig fails startup if empty while `idp.enabled=true`.
- pkg/config/cmd/controller: added audience field, wired into the existing authenticator audience check; logs a note at startup if left unset.

### User stories

- As an operator, enabling IDP auth with a minimal config should enforce role-based authorization by default.
- As an operator, an incomplete IDP config should fail at startup with a clear error.
- As an operator, I want to optionally restrict accepted tokens by audience.

### Documentation

Config-template updated (gateway/configs/config-template.toml).

### Samples

[controller.auth.idp]
roles_claim = "roles"   # required (defaulted) when enabled = true
audience = []           # optional; restricts accepted token audiences when set

### CODE REVIEW

File: gateway/gateway-controller/pkg/config/config.go [hunk 1]
```diff
@@ -1948,6 +1950,12 @@ func (c *Config) validateAuthConfig() error {
 		}
 	}
 
+	if c.Controller.Auth.IDP.Enabled && strings.TrimSpace(c.Controller.Auth.IDP.RolesClaim) == "" {
+		return fmt.Errorf("auth.idp.roles_claim is empty while auth.idp.enabled=true - " +
+			"Set auth.idp.roles_claim to the token claim carrying the caller's roles/scopes " +
+			"(and configure auth.idp.role_mapping to map them to local roles)")
+	}
```
[the remark this answers is not in the document]
Comment (thivindu) [reply gateway/gateway-controller/pkg/config/config.go:1957]: This is by design. An empty audience disables JWT audience validation for valid tokens.

File: gateway/gateway-controller/pkg/config/validator_test.go [hunk 2]
```diff
@@ -199,8 +199,9 @@ func TestValidateAuthConfig_IDPAuthEnabled(t *testing.T) {
 					Enabled: false,
 				},
 				IDP: IDPConfig{
-					Enabled: true,
-					JWKSURL: "https://idp.example.com/jwks",
+					Enabled:    true,
+					JWKSURL:    "https://idp.example.com/jwks",
+					RolesClaim: "roles",
```
[the remark this answers is not in the document]
Comment (thivindu) [reply gateway/gateway-controller/pkg/config/validator_test.go:204]: Audience was intentionally made warn-only (logged in main.go, not enforced in validateAuthConfig), per the chosen non-breaking design. An empty audience never causes validateAuthConfig to fail, so no fixture needs an Audience, and there is no config-time "audience validation error" to assert.

File: gateway/gateway-controller/pkg/config/config_test.go [hunk 3]
```diff
@@ -1504,20 +1504,23 @@ func TestConfig_ValidateAuthConfig(t *testing.T) {
...
 		errContains string
 	}{
 		{name: "IDP disabled", idpEnabled: false, wantErr: false},
-		{name: "IDP enabled no role mapping", idpEnabled: true, roleMapping: nil, wantErr: false},
-		{name: "IDP enabled single wildcard", idpEnabled: true, roleMapping: map[string][]string{"admin": {"*"}}, wantErr: false},
-		{name: "IDP enabled multiple wildcards", idpEnabled: true, roleMapping: map[string][]string{"admin": {"*"}, "user": {"*"}}, wantErr: true, errContains: "multiple wildcard ('*') mappings detected"},
+		{name: "IDP enabled no role mapping", idpEnabled: true, rolesClaim: "roles", roleMapping: nil, wantErr: false},
+		{name: "IDP enabled single wildcard", idpEnabled: true, rolesClaim: "roles", roleMapping: map[string][]string{"admin": {"*"}}, wantErr: false},
+		{name: "IDP enabled multiple wildcards", idpEnabled: true, rolesClaim: "roles", roleMapping: map[string][]string{"admin": {"*"}, "user": {"*"}}, wantErr: true, errContains: "multiple wildcard ('*') mappings detected"},
+		{name: "IDP enabled empty roles claim fails closed", idpEnabled: true, rolesClaim: "", roleMapping: nil, wantErr: true, errContains: "roles_claim is empty"},
```
[the remark this answers is not in the document]
Comment (thivindu) [reply gateway/gateway-controller/pkg/config/config_test.go:1516]: audience was intentionally made warn-only (logged in main.go, not enforced in validateAuthConfig), per the chosen non-breaking design. An empty audience never causes validateAuthConfig to fail, so no fixture needs an Audience, and there is no config-time "audience validation error" to assert.

File: common/authenticators/jwt_authenticator_test.go [hunk 4]
```diff
@@ -265,3 +265,67 @@ func TestJWTAuthenticator_Authenticate_ExpiredTokenRejected_WithIssuerValidation
...
+		"sub": "user123",
+		"exp": time.Now().Add(1 * time.Hour).Unix(),
+	})
+	tokenString, err := tok.SignedString(secret)
+	assert.NoError(t, err)
+
+	a := &JWTAuthenticator{
+		config: &models.AuthConfig{JWTConfig: idp},
+		logger: logger,
+		jwks:   staticKeyfunc{key: secret},
```
[the remark this answers is not in the document]
Comment (thivindu) [reply common/authenticators/jwt_authenticator_test.go:288]: This is outside of the current scope of the fix. This will be addressed later

File: gateway/gateway-controller/pkg/config/validator_test.go [hunk 5]
```diff
@@ -210,6 +211,28 @@ func TestValidateAuthConfig_IDPAuthEnabled(t *testing.T) {
 	assert.NoError(t, err)
 }
 
+func TestValidateAuthConfig_IDPAuthEnabledEmptyRolesClaim_FailsClosed(t *testing.T) {
+	// IDP enabled with an empty roles_claim silently disables authorization
+	// (the JWT authenticator sets skip-authz, bypassing the per-route role
+	// check), granting every valid token full admin access. Refuse to start.
```
[the remark this answers is not in the document]
Comment (thivindu) [reply gateway/gateway-controller/pkg/config/validator_test.go]: Addressed via - https://github.com/wso2/api-platform/pull/3012/changes/c746ed68cf244538af9b6c04cf17dd482fba120e

File: common/authenticators/jwt_authenticator_test.go [hunk 6]
```diff
@@ -265,3 +265,67 @@ func TestJWTAuthenticator_Authenticate_ExpiredTokenRejected_WithIssuerValidation
...
+	}
+	req, _ := http.NewRequest("GET", "/", nil)
+	req.Header.Set(constants.AuthorizationHeader, constants.BearerPrefix+tokenString)
+
+	result, err := a.Authenticate(req)
+	assert.NoError(t, err)
+	return result
```
[the remark this answers is not in the document]
Comment (thivindu) [reply common/authenticators/jwt_authenticator_test.go:295]: Addressed via - https://github.com/wso2/api-platform/pull/3012/changes/c746ed68cf244538af9b6c04cf17dd482fba120e

**Comment by thivindu** [comment]
@coderabbitai review

**Comment by thivindu** [comment]
@coderabbitai review

**Comment by thivindu** [comment]
@coderabbitai approve

File: common/authenticators/jwt_authenticator_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply common/authenticators/jwt_authenticator_test.go:288]: **Fix:**
- Add `jwt.WithValidMethods([]string{"RS256", "PS256", "ES256"})` to the `ParseWithClaims` call in `Authenticate` — same set as the JWT auth policy
- Hardcode this list; it's a security invariant (blocks `HS256`/`none`), not operator-configurable — the JWKS already communicates the IDP's actual algorithm
- Update the test helper to sign with an RSA/EC key instead of `[]byte("test-secret")` + `HS256`, and add negative-path tests confirming HMAC and `none` tokens are rejected

**Review by renuka-fernando — APPROVED** [review]
